### ISSUE TITLE
How to create JsonSerializable to send a POST with body

### ISSUE THREAD

**Issue by brettpappas** [Clancey/simple_auth#48]
I have tried a number of ways to create an API definition that sends a POST request with the parameters as part of the body.

Dropbox requires calls to it's API to send it's parameters as JSON body in a post request. I tried creating a definition but I don't know how to pass the body:

```
@DropboxApiDeclaration("DropboxTestApi", "<client_id>", "<client_secret>", "<redirect>")
abstract class DropboxTestDefinition {
    @Post(url: "https://api.dropboxapi.com/2/files/list_folder")
    Future<Response<Object>> getFileList();
}
```

Any chance you can provide an example?

**Comment by Clancey** [comment]
Yes, there are two ways you can do this. One is by using the [generator](https://github.com/Clancey/simple_auth/tree/master/simple_auth_generator). That is how you would use the code you linked in the issue. A more complete example would be the [youtube](https://github.com/Clancey/simple_auth/blob/master/simple_auth_flutter_example/lib/api_definitions/youtubeApi.dart#L19-L23). It doesn't have post but you just annotate a property with [body](https://github.com/Clancey/simple_auth/blob/master/simple_auth/lib/src/annotations.dart#L252).

The other option is to manually build your requests. There is a sample of that [here](https://github.com/Clancey/simple_auth/blob/master/simple_auth/lib/src/providers/google.dart#L40-L46).

**Comment by brettpappas** [comment]
Thanks for the response. I did try both of these options and still couldn't get the body passed to the request. I will try debugging the code to see where it is getting lost.

**Comment by brettpappas** [comment]
@Clancey The problem seems to be with the `replace` method in the [Request](https://github.com/Clancey/simple_auth/blob/master/simple_auth/lib/src/request.dart#L36-L56) object, whenever this is called - like when it wants to set the oauth token in the headers [(authenticateRequest)](https://github.com/Clancey/simple_auth/blob/master/simple_auth/lib/src/oauth/oauthApi.dart#L100-L106) - it will remove the `body` property.

A simple solution would be to add the body property to the replace method. Another option would be to use replaceBody method everywhere instead of replace but that seems less ideal.

Do you want me to submit a PR for this or do you have a better solution?

**Comment by Clancey** [comment]
I had an issue with this when Dart2 came out: https://github.com/Clancey/simple_auth/commit/e9917fc5811de4603e26ac9ffa362252a6cd4603

I think I may have botched it then. Adding it to the replace method.
